arrow-continuations/src/main/kotlin/arrow/continuations/Reset.kt

```diff
@@ -0,0 +1,37 @@
+package arrow.continuations
+
+import arrow.continuations.generic.DelimContScope
+import arrow.continuations.generic.DelimitedScope
+import arrow.continuations.generic.NestedDelimContScope
+
+object Reset {
+  /**
+   * Implements delimited continuations with no multi-shot support.
+   *
+   * For a version that simulates multi-shot (albeit with drawbacks) see [multi].
+   * For a version that allows nesting [reset] and calling parent scopes inside inner scopes see [nested].
+   */
+  suspend fun <A> single(block: suspend DelimitedScope<A>.() -> A): A =
+    DelimContScope.reset(block)
+
+  /**
+   * (Simulated) Multi-shot capable delimited control scope
+   * [block] will rerun completely and only the results of shift are cached.
+   * All side-effects outside of shift will rerun on each shift callback invocation
+   *
+   * [multi] is a synthetic function that requires compiling with the arrow plugin.
+   * The Arrow plugin ensures bindings are ordered and top-level before effects
+   * and pure expressions to avoid side-effects rerunning in multi-shot mode
```

Comment:
This is an interesting approach to the problem of rerunning. I wonder how this will turn out :thinking: 

Another idea would be the manual CPS transformation for every method that does multishot, to avoid using a `Continuation` but rather a function/closure directly. But that comes with challenges of its own.